20240808-custom-limit-email-notifications.md

```diff
@@ -0,0 +1,194 @@
+
+The chore would:
+
+* for each project on the satellite:
+    * if project does not have any custom limits set, skip
+    * query current project usage
+    * for each custom limit type (storage, egress):
+        * if project does not have custom limit for this type, skip
+        * for each threshold (80%, 100%):
+             * if project usage for this limit type has exceeded threshold:
+                 * if email was already sent for this "threshold event", skip
+                 * send email for this threshold event, and update project flags in DB to indicate an email was sent
+            * if project usage for this limit type has *not* exceeded threshold:
+                * if email was set for this threshold event, reset the value to indicate "unsent" (this means that we have entered a new billing period or the user has reset their custom limits)
+
+The chore could have multiple "workers" to allow parallelization, and how often the chore runs would be configurable.
+
+Pros:
+
+* Only requires adding "email sent" flags to `projects` DB table - simple schema change
+* The entire feature being on the satellite core avoids race conditions that are introduced by depending on metainfo/API pods
+
+Cons:
+
+* Timeliness of emails depends on how often the chore runs - see "alternatives considered" for more info about why we did not choose a more "responsive" method
+
+#### Summary
+
+* Updating projects table schema to have flags indicating when emails were sent
+* Adding a chore on the satellite core to detect when emails need to be sent, send emails, and keep flags pertaining to project notifications in the DB up to date
+
+## Disclaimers
+
+### Anti-goals
+
+* impacting metainfo performance
+* sending excessive limit notifications (only one email for 80% and one for 100%, per limit per project)
+* failing to send a limit notification when the user passes the 80% or 100% threshold
+* sending emails to users who have not set custom limits
+
+### Alternatives considered
+
+1. Use an existing chore on the satellite core rather than adding a new one
+    * tally chore is a possible candidate
+    * mixes up email notification logic with other features
+2. Use third-party tools, and send emails completely outside the satellite
+    * it would not use the same email sender as other satellite emails to the user
+    * could run on non-prod DB, which wouldn't impact satellite, but would introduce some delay (however often the backup gets updated)
+    * not dependent on satellite release cycle (quicker to set up/test/adjust)
+    * more access/specialized knowledge required for maintenance - satellite code is more "accessible"
+    * likely requires two separate third-parties - one to do queries, and one to send emails. If we go with existing tools for these, the task my require cross-team work with datascience and/or marketing.
+3. Detect "threshold events" during metainfo validation, at the time of usage
```

Comment:
TODO: consider making this the idea for the main design. If it is important for users to receive emails asap after an event occurs, we might need to detect events during metainfo validation